File: clase-08-19/tads/LinkedQueue.cpp
#pragma once

#include "Queue.cpp"
#include <assert.h>

template <class K>
struct LinkedQNode
{
  K data;
  LinkedQNode<K> *next = nullptr;
};

template <class T>
class LinkedQueue : public Queue<T>
{
private:
  int elements = 0;
  LinkedQNode<T> *front = nullptr;
  LinkedQNode<T> *back = nullptr;

public:
  LinkedQueue()
  {
  }

  ~LinkedQueue()
  {
    while (!isEmpty())
    {
      pop();
    }
  }

  int size() override
  {
    return elements;
  }

  void push(T data) override
  {
    LinkedQNode<T> *newNode = new LinkedQNode<T>();
    newNode->data = data;
    if (isEmpty())
    {
      front = newNode;
      back = newNode;
    }
    else
    {
      back->next = newNode;
      back = back->next;
    }

    elements++;
  }

  T top() override
  {
    assert(!isEmpty());
    return front->data;
  }

  void pop() override
  {
    if (isEmpty())
    {
      return;
    }

    LinkedQNode<T> *temp = front;
    front = front->next;
    delete temp;
    if (elements == 1)
    {
      back = nullptr;
    }
    elements--;
  }

  bool isFull() override
  {
    return false;
  }

  bool isEmpty() override
  {
    return elements == 0;
  }
};

File: clase-08-19/tads/LinkedQueue.java
package tads;

public class LinkedQueue<T> implements Queue<T> {
  class Node {
    T data;
    Node next;
  }

  private int elements = 0;
  private Node front = null;
  private Node back = null;

  @Override
  public int size() {
    return elements;
  }

  @Override
  public T top() {
    if (isEmpty()) {
      throw new RuntimeException("top on empty queue");
    }

    return front.data;
  }

  @Override
  public void push(T data) {
    Node newNode = new Node();
    newNode.data = data;
    if (isEmpty()) {
      front = newNode;
      back = newNode;
    } else {
      back.next = newNode;
      back = back.next;
    }

    elements++;
  }

  @Override
  public void pop() {
    if (isEmpty()) {
      return;
    }

    front = front.next;
    if (elements == 1) {
      back = null;
    }
    elements--;
  }

  @Override
  public boolean isFull() {
    return false;
  }

  @Override
  public boolean isEmpty() {
    return elements == 0;
  }

}
